src/csharp/Microsoft.Spark/Sql/Types/ComplexTypes.cs

```diff
@@ -10,10 +10,20 @@
 
 namespace Microsoft.Spark.Sql.Types
 {
+    /// <summary>
+    /// Complex types
+    /// </summary>
+    public abstract class ComplexType : DataType
+    {
+        internal override bool NeedConversion() => true;
+
+        internal override object FromInternal(object obj) => throw new NotImplementedException();
+    }
+
     /// <summary>
     /// An array type containing multiple values of a type.
     /// </summary>
-    public sealed class ArrayType : DataType
+    public sealed class ArrayType : ComplexType
```

Comment:
Since these types are public, we wanted to match the Scala types, thus DataType.

Reply:
Reverted.